### PR TITLE
Add TypeScript type definitions

### LINKED ISSUE #21: Bring typescript type definitions into this this repo
PEG.js type definitions currently live in DefinitelyTyped - https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/pegjs

we should bring these into the peggy repo to prevent people from having to install the type definitions separately and to avoid the separate DefinitelyTyped workflow for contributors.

### PR DESCRIPTION
**Pull request by phpnode**
Also cleans up the files array in package.json in favour of .npmignore

Fixes #21

### CODE REVIEW

**Comment by phpnode** [comment]
output of `npm pack --dry-run` after these changes:
```
npm notice 
npm notice 📦  peggy@0.11.0
npm notice === Tarball Contents === 
npm notice 1.5kB   AUTHORS                                          
npm notice 21.0kB  CHANGELOG.md                                     
npm notice 2.2kB   CONTRIBUTING.md                                  
npm notice 1.1kB   LICENSE                                          
npm notice 20.8kB  README.md                                        
npm notice 8.8kB   bin/peggy                                        
npm notice 1.6kB   lib/compiler/asts.js                             
npm notice 2.7kB   lib/compiler/index.js                            
npm notice 2.3kB   lib/compiler/js.js                               
npm notice 1.5kB   lib/compiler/opcodes.js                          
npm notice 14.9kB  lib/compiler/passes/generate-bytecode.js         
npm notice 42.8kB  lib/compiler/passes/generate-js.js               
npm notice 859B    lib/compiler/passes/remove-proxy-rules.js        
npm notice 1.5kB   lib/compiler/passes/report-duplicate-labels.js   
npm notice 691B    lib/compiler/passes/report-duplicate-rules.js    
npm notice 1.3kB   lib/compiler/passes/report-infinite-recursion.js 
npm notice 996B    lib/compiler/passes/report-infinite-repetition.js
npm notice 510B    lib/compiler/passes/report-undefined-rules.js    
npm notice 1.9kB   lib/compiler/visitor.js                          
npm notice 360B    lib/grammar-error.js                             
npm notice 164.4kB lib/parser.js                                    
npm notice 6.4kB   lib/peg.d.ts                                     
npm notice 1.4kB   lib/peg.js                                       
npm notice 1.4kB   package.json                                     
npm notice === Tarball Details === 
npm notice name:          peggy                                   
npm notice version:       0.11.0                                  
npm notice filename:      peggy-0.11.0.tgz                        
npm notice package size:  60.3 kB                                 
npm notice unpacked size: 302.9 kB                                
npm notice shasum:        b898eeb4494e4d674c87e4b4a744aa86dfdae826
npm notice integrity:     sha512-7b1wD+6kD9gjI[...]ppRwvH7XyvWtQ==
npm notice total files:   24                                      
npm notice 
peggy-0.11.0.tgz
```

File: .npmignore [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+.github
+benchmark
+bin/.eslintrc.json
+browser
+docs
+examples
+lib/.eslintrc.json
+src
+test
+tools
```
Comment (hildjj) [inline .npmignore:12]: I don't think CHANGELOG.md should ship, but I'm willing to be talked out of that.  I dislike that the .eslintrc's in the subdirs are in different format than the base one, but we can clean that up later.

File: lib/peg.d.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  name: string;
+  message: string;
+  location: LocationRange;
+  found?: any;
+  expected?: ExpectedItem[];
+  stack?: any;
+}
+
+// for backwards compatibility with PEGjs
+export interface PegjsError extends PeggyError {}
```
Comment (hildjj) [inline lib/peg.d.ts:48]: How do we see the backward-compatibility bits being used?

File: lib/peg.d.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+export type GrammarError = PeggyError;
+export var GrammarError: any;
+
+export interface ParserOptions {
+  startRule?: string;
+  tracer?: any;
```
Comment (hildjj) [inline lib/peg.d.ts:55]: `tracer` is almost undocumented. We can leave fixing this as a TODO, but let's file a post-1.0 bug.

File: lib/peg.d.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  optimize?: "speed" | "size";
+  /** plugins to use */
+  plugins?: any[];
+  /** makes the parser trace its progress (default: `false`) */
+  trace?: boolean;
+}
+
+export interface ParserBuildOptions extends BuildOptionsBase {
+  /** if set to `"parser"`, the method will return generated parser object; if set to `"source"`, it will return parser source code as a string (default: `"parser"`) */
+  output?: "parser";
```
Comment (hildjj) [inline lib/peg.d.ts:80]: shouldn't this be `"parser" | "source"`?

Comment (hildjj) [reply lib/peg.d.ts:80]: oh, nm, i see the other overloads below

File: lib/peg.d.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+export interface OutputFormatAmdCommonjs extends BuildOptionsBase {
+  /** if set to `"parser"`, the method will return generated parser object; if set to `"source"`, it will return parser source code as a string (default: `"parser"`) */
+  output: "source";
+  /** format of the genreated parser (`"amd"`, `"bare"`, `"commonjs"`, `"globals"`, or `"umd"`); valid only when `output` is set to `"source"` (default: `"bare"`) */
+  format: "amd" | "commonjs";
```
Comment (hildjj) [inline lib/peg.d.ts:87]: need to add "es" to all of these

**Review by hildjj — CHANGES_REQUESTED** [review]
I think the "es" addition is the only real blocking change needed.

File: lib/peg.d.ts (hunk 2 above)
Comment (phpnode) [reply lib/peg.d.ts:48]: I think it's conceivable that someone is doing `import { PegjsError } from "pegjs"` so didn't want to just rename it, and this was the only type that includes the old name.

Comment (phpnode) [reply lib/peg.d.ts:48]: I will change it to just a type alias though, no need to extend

File: lib/peg.d.ts (hunk 3 above)
Comment (hildjj) [reply lib/peg.d.ts:55]: See #24

**Review by hildjj — APPROVED** [review]

File: lib/peg.d.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  interface LocationRange {
+    start: Location;
+    end: Location;
+  }
+
+  class SyntaxError {
+    line: number;
+    column: number;
+    offset: number;
+    location: LocationRange;
```
Comment (Mingun) [inline lib/peg.d.ts:21]: This is quite strange... `location` should be enough. Need to check if the `line/column/offset` are really filled in the error

Comment (phpnode) [reply lib/peg.d.ts:21]: it might need revisiting, I've just copied the definitions from DefinitelyTyped here

Comment (Mingun) [reply lib/peg.d.ts:21]: Just checked this, `line/column/offset` is definedly an error:
https://github.com/peggyjs/peggy/blob/2a37f1f02b57a3c5a2277d2735b2773de30d2aa4/lib/compiler/passes/generate-js.js#L754-L764

Comment (hildjj) [reply lib/peg.d.ts:21]: @mingun can you file a new bug please?  I'll mark it release-blocking

Comment (Mingun) [reply lib/peg.d.ts:21]: #29

**Pull request merged**